### ISSUE TITLE
making new types fails when config setting attributes.allow-all=false

### ISSUE THREAD

**Issue by galdosd** [thinkaurelius/titan#611]
When disabling attributes.allow-all, making new labels/keys fails. I guess titan itself is using the default serializer for storing type metadata? FWIW I am using cassandra backend, clearing the keyspace each time before tests, and testing by adding just a single String-typed attribute.

This is not a huge deal -- we can just keep allow-all=true whenever loading the schema (should be a relatively rare occurrence in production) and then disable it and reconnect, but it would be nice to document this gotcha in the same places where attributes.allow-all usage is introduced (I got the idea from https://github.com/thinkaurelius/titan/wiki/Datatype-and-Attribute-Serializer-Configuration)

``` java
com.thinkaurelius.titan.core.TitanException: Could not commit transaction due to exception during persistence
        at com.thinkaurelius.titan.graphdb.transaction.StandardTitanTx.commit(StandardTitanTx.java:1029)
        at com.thinkaurelius.titan.graphdb.blueprints.TitanBlueprintsGraph.commit(TitanBlueprintsGraph.java:57)
        ... 30 more
Caused by: com.thinkaurelius.titan.core.TitanException: Unexpected exception during backend operation
        at com.thinkaurelius.titan.diskstorage.util.BackendOperation.execute(BackendOperation.java:67)
        at com.thinkaurelius.titan.graphdb.database.StandardTitanGraph.commit(StandardTitanGraph.java:320)
        at com.thinkaurelius.titan.graphdb.transaction.StandardTitanTx.commit(StandardTitanTx.java:1018)
        ... 29 more
Caused by: com.esotericsoftware.kryo.KryoException: java.lang.IllegalArgumentException: Class is not registered: com.thinkaurelius.titan.core.Order
Note: To register this class use: kryo.register(com.thinkaurelius.titan.core.Order.class);
Serialization trace:
value (com.thinkaurelius.titan.graphdb.types.TypeAttribute)
        at com.esotericsoftware.kryo.serializers.FieldSerializer$ObjectField.write(FieldSerializer.java:585)
        at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:213)
        at com.esotericsoftware.kryo.Kryo.writeObject(Kryo.java:486)
        at com.thinkaurelius.titan.graphdb.database.serialize.kryo.KryoDataOutput.writeObjectNotNull(KryoDataOutput.java:81)
        at com.thinkaurelius.titan.graphdb.database.EdgeSerializer.writeRelation(EdgeSerializer.java:342)
        at com.thinkaurelius.titan.graphdb.database.StandardTitanGraph.persist(StandardTitanGraph.java:345)
        at com.thinkaurelius.titan.graphdb.database.StandardTitanGraph.access$000(StandardTitanGraph.java:49)
        at com.thinkaurelius.titan.graphdb.database.StandardTitanGraph$2.call(StandardTitanGraph.java:301)
        at com.thinkaurelius.titan.graphdb.database.StandardTitanGraph$2.call(StandardTitanGraph.java:241)
        at com.thinkaurelius.titan.diskstorage.util.BackendOperation.execute(BackendOperation.java:62)
        ... 31 more
Caused by: java.lang.IllegalArgumentException: Class is not registered: com.thinkaurelius.titan.core.Order
Note: To register this class use: kryo.register(com.thinkaurelius.titan.core.Order.class);
        at com.esotericsoftware.kryo.Kryo.getRegistration(Kryo.java:442)
        at com.esotericsoftware.kryo.util.DefaultClassResolver.writeClass(DefaultClassResolver.java:79)
        at com.esotericsoftware.kryo.Kryo.writeClass(Kryo.java:472)
        at com.esotericsoftware.kryo.serializers.FieldSerializer$ObjectField.write(FieldSerializer.java:561)
        ... 40 more

```

Versions:

``` xml
        <dependency>
           <groupId>com.thinkaurelius.titan</groupId>
           <artifactId>titan-core</artifactId>
           <version>0.4.2</version>
        </dependency>
        <dependency>
           <groupId>com.thinkaurelius.titan</groupId>
           <artifactId>titan-cassandra</artifactId>
           <version>0.4.2</version>
        </dependency>
```

**Comment by mbroecheler** [comment]
Thanks for pointing this out. This has been fixed in the next release.

On Fri, Mar 14, 2014 at 6:59 PM, Domingo Ignacio Galdos <
notifications@github.com> wrote:

> When disabling attributes.allow-all, making new labels/keys fails. I guess
> titan itself is using the default serializer for storing type metadata?
> FWIW I am using cassandra backend, clearing the keyspace each time before
> tests, and testing by adding just a single String-typed attribute.
> 
> This is not a huge deal -- we can just keep allow-all=true whenever
> loading the schema (should be a relatively rare occurrence in production)
> and then disable it and reconnect, but it would be nice to document it in
> the same places where attributes.allow-all usage is mentioned (I got the
> idea from
> https://github.com/thinkaurelius/titan/wiki/Datatype-and-Attribute-Serializer-Configuration
> )
> 
> com.thinkaurelius.titan.core.TitanException: Could not commit transaction due to exception during persistence
>         at com.thinkaurelius.titan.graphdb.transaction.StandardTitanTx.commit(StandardTitanTx.java:1029)
>         at com.thinkaurelius.titan.graphdb.blueprints.TitanBlueprintsGraph.commit(TitanBlueprintsGraph.java:57)
>         ... 30 moreCaused by: com.thinkaurelius.titan.core.TitanException: Unexpected exception during backend operation
>         at com.thinkaurelius.titan.diskstorage.util.BackendOperation.execute(BackendOperation.java:67)
>         at com.thinkaurelius.titan.graphdb.database.StandardTitanGraph.commit(StandardTitanGraph.java:320)
>         at com.thinkaurelius.titan.graphdb.transaction.StandardTitanTx.commit(StandardTitanTx.java:1018)
>         ... 29 moreCaused by: com.esotericsoftware.kryo.KryoException: java.lang.IllegalArgumentException: Class is not registered: com.thinkaurelius.titan.core.OrderNote: To register this class use: kryo.register(com.thinkaurelius.titan.core.Order.class);Serialization trace:value (com.thinkaurelius.titan.graphdb.types.TypeAttribute)
>         at com.esotericsoftware.kryo.serializers.FieldSerializer$ObjectField.write(FieldSerializer.java:585)
>         at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:213)
>         at com.esotericsoftware.kryo.Kryo.writeObject(Kryo.java:486)
>         at com.thinkaurelius.titan.graphdb.database.serialize.kryo.KryoDataOutput.writeObjectNotNull(KryoDataOutput.java:81)
>         at com.thinkaurelius.titan.graphdb.database.EdgeSerializer.writeRelation(EdgeSerializer.java:342)
>         at com.thinkaurelius.titan.graphdb.database.StandardTitanGraph.persist(StandardTitanGraph.java:345)
>         at com.thinkaurelius.titan.graphdb.database.StandardTitanGraph.access$000(StandardTitanGraph.java:49)
>         at com.thinkaurelius.titan.graphdb.database.StandardTitanGraph$2.call(StandardTitanGraph.java:301)
>         at com.thinkaurelius.titan.graphdb.database.StandardTitanGraph$2.call(StandardTitanGraph.java:241)
>         at com.thinkaurelius.titan.diskstorage.util.BackendOperation.execute(BackendOperation.java:62)
>         ... 31 moreCaused by: java.lang.IllegalArgumentException: Class is not registered: com.thinkaurelius.titan.core.OrderNote: To register this class use: kryo.register(com.thinkaurelius.titan.core.Order.class);
>         at com.esotericsoftware.kryo.Kryo.getRegistration(Kryo.java:442)
>         at com.esotericsoftware.kryo.util.DefaultClassResolver.writeClass(DefaultClassResolver.java:79)
>         at com.esotericsoftware.kryo.Kryo.writeClass(Kryo.java:472)
>         at com.esotericsoftware.kryo.serializers.FieldSerializer$ObjectField.write(FieldSerializer.java:561)
>         ... 40 more
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/thinkaurelius/titan/issues/611
> .

## 

Matthias Broecheler
http://www.matthiasb.com

**Comment by galdosd** [comment]
Fantastic, thank you very much for your time! I'll try it out as soon as the release is out; I'm a big fan of any and all safety-adding features.

**Comment by crakjie** [comment]
I have the same error with titan 0.4.4, is it corrected in this version?

**Comment by mbroecheler** [comment]
This has been fixed in 0.5 but (if my memory serves me correctly) a fix in
0.4 would have been backwards incompatible and hence we didn't do it.

On Thu, Oct 2, 2014 at 5:54 AM, etienne notifications@github.com wrote:

> I have the same error with titan 0.4.4, is it corrected in this version?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/thinkaurelius/titan/issues/611#issuecomment-57625046.

## 

Matthias Broecheler
http://www.matthiasb.com
